csrc/ops/alias.cpp

```diff
@@ -691,12 +691,18 @@ TensorView* slice(TensorView* inp, const std::vector<Slice>& ranges) {
       out_root_id = inp_root_id->cloneWithoutRFactor();
       out_rf_id = out_root_id;
     } else {
+      // Clip the start and stop values to the extent of the input
+      auto clipped_start =
```

Comment:
What about negative values? Should they be clipped to zero?

Reply:
Ordinarily we would add extent to negative values, which we can do with a where expression.

Reply:
What would happen if it's still negative?

Reply:
If it's negative at runtime we will wind up padding I think, which is also incorrect. I pushed a change just now that adds `where()` to handle negative start/stop. However, it leads to an error when trying to simplify the resulting index expressions. We can handle constant negative values by translating them at definition time, but that would still mean that dynamic `Val`s that take on negative values would lead to incorrect outputs.

Reply:
Example output:
```
%kernel_math {
b5 = 11 < 0;
i4 = 11 + 9;
i6 = where(b5, i4, 11);
b8 = 13 < 0;
i7 = 13 + 9;
i9 = where(b8, i7, 13);
T1_g[ iS2{( ( -( fmin(9, ( where(( 11 < 0 ), ( 11 + 9 ), 11) )) ) ) + ( fmin(9, ( where(( 13 < 0 ), ( 13 + 9 ), 13) )) ) )}rf ]
   = slice( T0_g[ iS0{9} ], { {i6, i9, 1} } )
}
```

Reply:
> What would happen if it's still negative?

I will check in the AM but I think in pytorch it's an error to provide a value < -extent.

Reply:
In both torch and numpy, it seems that providing a negative value first adds `extent` then clips to `>= 0`. Values are also clipped to `<= extent`, so any int is valid input for either start or stop and the expressions to use for `a` equal to either start or stop are:
```python
normed_a = min(max(where(a < 0, extent + a, a), 0), extent)
```
These will be pretty tough to simplify if we don't enforce which branch we're on. We can do that at concretization by determining the branch as one of:
```c++
class enum SliceIndexBranch {
  AlwaysZero, // a <= -extent
  Negative, // -ext < a < 0
  Positive, // 0 <= a < extent
  AlwaysExtent // extent <= a
};
```
These map to the following much simpler expressions
```
AlwaysZero => zeroVal(),
Negative => extent() + a,
Positive => a,
AlwaysExtent => extent()
```
I think doing this at concretization makes sense because users commonly write slicing code with only one or two of these branches in mind, so we should not try and compile a kernel that handles them all. For example, a user might hardcode an upper limit that could be either `Positive` or `AlwaysExtent`, but they would not ordinarily mix `Negative` and `Positive` in the same code I would think. Even if they do, by concretizing we'll have faster kernels, and the only price paid is a recompile when we see a new pattern.